File: KSmallestin BST.java
// Time Complexity : O(N)
// Space Complexity : O(H)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int k;
    int result = -1;
    private void inorder(TreeNode root){
        if(root == null || this.k == 0) return;
        inorder(root.left);
        k--;
        if(k == 0){
            result = root.val;
            return;    
        }
        inorder(root.right);
        
    }
    public int kthSmallest(TreeNode root, int k) {
        this.k = k;
        if(root == null) return -1;
        inorder(root);
        return result;
        
        // if(root == null) return -1;
        // Stack<TreeNode> s = new Stack<>();
        // while(!s.isEmpty() || root != null){
        //     while(root!=null){
        //         s.push(root);
        //         root = root.left;
        //     }
        //     root = s.pop();
        //     k--;
        //     if(k == 0) return root.val;
        //     root = root.right;
        // }
        // return -1;
        
    }
}

File: k_smallest_in_bst.py
# // Time Complexity : O(n)
# // Space Complexity : O(h)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : Nope


# // Your code here along with comments explaining your approach

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

############Iterative Solution ###############
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        n=0
        stack=[]
        curr=root

        #Until we reach the end or the stack is non empty
        while curr or stack:
            #whle we reach the leaf
            while curr:
                #Add the elemtns to the stack
                stack.append(curr)
                #Move to the left
                curr=curr.left
            
            #When we reach the leaf remove the top of the stack
            curr=stack.pop()
            #increase the n
            n+=1

            #When n==k return the curr value
            if n==k:
                return curr.val

            #Move to the right
            curr=curr.right